content/docs/hoon/reference/rune/sig.md

```diff
@@ -314,139 +591,277 @@ From the kernel:
   $(a (dec a), b +(b))
 ```
 
-### `~&` "sigpam"
+---
+
+## `~&` "sigpam"
+
+Debugging printf.
+
+#### Syntax
+
+Two arguments, fixed.
+
+<table>
+<tr><th>Form</th><th>Syntax</th></tr>
+<tr>
+<td>Tall</td>
+<td>
+<pre>
+~&  p
+q
+</pre>
+</td>
+</tr>
+<tr>
+<td>Wide</td>
+<td>
+<pre>
+~&(p q)
+</pre>
+</td>
+</tr>
+<tr>
+<td>Irregular</td>
+<td>None.</td>
+</tr>
+</table>
+
+#### AST
 
-`[%sgpm p=hoon q=hoon]`: debugging printf.
+```hoon
+[%sgpm p=hoon q=hoon]
+```
 
-##### Expands to
+#### Expands to
 
 `q`.
 
-##### Product
+#### Product
 
 Pretty-prints `p` on the console before computing `q`.
 
-##### Discussion
+#### Discussion
 
-This rune has no semantic effect beyond the Hoon expression `q`.  It's used solely to create a side-effect: printing the value of `p` to the console.
+This rune has no semantic effect beyond the Hoon expression `q`. It's used
```

Comment:
I'm not aware of a type issue with `~&`

Reply:
oh you mean how if eg you return either a @ud or @t it always prints an @t. This isn't to do with `~&`, it's because any kind of conditional fork expression eg:

```
?:  ...
  12
 'foo'
 ```
 
 is evaluated to return a union of its possible types, in this case `?(@ud @t)`. The problem is that `$?` can't differentiate between auras so it just casts everything to the default which is the last item in the union, a `@t` in this case.
 
 so the problem isn't `~&`, the problem is the limitations of the type system. You'd have the same problem in any case where you were returning a union of auras